feat(ndt_scan_matcher): estimate NDT covariance in real-time with limited initial positions

KOKIAOKI opened the pull request:
## Description
Calculate 2D covariance (xx, xy, yx, yy) in real time using the convergence of NDT.
I limit the arrangement of multiple initial positions based on the Hessian matrix of the NDT score function.
## Related links
[original paper](https://www.fujipress.jp/jrm/rb/robot003500020435/)
## Tests performed
I evaluated it In a tunnel evaluation (already shared in weekly meeting).
The real-time covariance estimation obtained the same convergence tendency as offline convergence evaluation.
## Notes for reviewers
Related PRs:
- https://github.com/tier4/ndt_omp/pull/22
Output Hessian matrix
- https://github.com/autowarefoundation/autoware_launch/pull/643
Parameters for this covariance estaimtion
## Interface changes
- topic
`/localization/pose_estimator/multi_ndt_pose`
 ->NDT convergence positions
`/localization/pose_estimator/multi_initial_pose`
 ->NDT initial positions

-  yaml setting (ndt_scan_matcher.param.yaml)
Category: 2D Real-time covariance estimation (output_pose_covariance is the minimum value)
Default use_covariance_estimation is **false**.

## Effects on system behavior
<!-- Describe how this PR affects the system behavior. -->
## Pre-review checklist for the PR author
The PR author **must** check the checkboxes below when creating the PR.
- [x] I've confirmed the [contribution guidelines].
- [x] The PR follows the [pull request guidelines].
## In-review checklist for the PR reviewers
The PR reviewers **must** check the checkboxes below before approval.
- [ ] The PR follows the [pull request guidelines].
- [ ] The PR has been properly tested.
- [ ] The PR has been reviewed by the code owners.
## Post-review checklist for the PR author
The PR author **must** check the checkboxes below before merging.
- [ ] There are no open discussions or they are tracked via tickets.
- [ ] The PR is ready for merge.
After all checkboxes are checked, anyone who has write access can merge the PR.
[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

KYabuuchi commented on localization/ndt_scan_matcher/src/ndt_scan_matcher_core.cpp:
```diff
@@ -704,6 +723,74 @@ bool NDTScanMatcher::validate_converged_param(
+    rot = Eigen::Rotation2Dd(th);
+  }
+
+  // first result is added to mean and covariance
+  Eigen::Vector2d tmp_centroid;
+  Eigen::Matrix2d tmp_cov;
+  Eigen::Vector2d p2d(ndt_result.pose(0, 3), ndt_result.pose(1, 3));
+  tmp_centroid += p2d;
+  tmp_cov += p2d * p2d.transpose();
```
`tmp_centroid` and `tmp_cov` are not zero-initialized. Therefore `tmp_centroid += p2d;` may be unintended value.

```suggestion
  // first result is added to mean and covariance
  Eigen::Vector2d p2d(ndt_result.pose(0, 3), ndt_result.pose(1, 3));
  Eigen::Vector2d tmp_centroid = p2d;
  Eigen::Matrix2d tmp_cov = p2d * p2d.transpose();
```

KOKIAOKI replied:
Thank you for checking my code. I reflect your corrections.

KYabuuchi reviewed (commented):
Thank you for creating the PR. I found a risky code in a brief review. Please check.

kminoda commented on localization/ndt_scan_matcher/src/ndt_scan_matcher_core.cpp:
```diff
@@ -141,6 +141,15 @@ NDTScanMatcher::NDTScanMatcher()
   initial_pose_distance_tolerance_m_ =
     this->declare_parameter<double>("initial_pose_distance_tolerance_m");
 
+  use_cov_estimation_ = this->declare_parameter("use_covariance_estimation", use_cov_estimation_);
```
Please do not use the default parameter and instead do as follow
```suggestion
  use_cov_estimation_ = this->declare_parameter<bool>("use_covariance_estimation");
```

kminoda commented on localization/ndt_scan_matcher/src/ndt_scan_matcher_core.cpp:
```diff
@@ -478,6 +490,13 @@ void NDTScanMatcher::callback_sensor_points(
     RCLCPP_WARN(get_logger(), "Not Converged");
   }
 
+  // covariance estimation
+  std::array<double, 36> ndt_covariance;
+  ndt_covariance = output_pose_covariance_;
+  if (is_converged && use_cov_estimation_) {
+    estimate_covariance(ndt_result, initial_pose_matrix, ndt_covariance, sensor_ros_time);
+  }
```
I believe this part is relatively computationally heavy, so I would recommend to move this right after the following part:
```
  // perform ndt scan matching
  (*state_ptr_)["state"] = "Aligning";
  const Eigen::Matrix4f initial_pose_matrix =
    pose_to_matrix4f(interpolator.get_current_pose().pose.pose);
  auto output_cloud = std::make_shared<pcl::PointCloud<PointSource>>();
  ndt_ptr_->align(*output_cloud, initial_pose_matrix);
  const pclomp::NdtResult ndt_result = ndt_ptr_->getResult();
  (*state_ptr_)["state"] = "Sleeping
```

This will also enable users to monitor the total execution time with `exe_time` as well.

kminoda commented on localization/ndt_scan_matcher/src/ndt_scan_matcher_core.cpp:
```diff
@@ -704,6 +723,72 @@ bool NDTScanMatcher::validate_converged_param(
+  // Laplace approximation
+  const Eigen::Matrix2d hessian_inv = -ndt_result.hessian.inverse().block(0, 0, 2, 2);
+  const Eigen::SelfAdjointEigenSolver<Eigen::Matrix2d> lap_eigensolver(hessian_inv);
+  Eigen::Matrix2d rot;
+  rot = Eigen::Rotation2Dd(0.0);
```
Please initialize the value at the same time with the declaration.

kminoda commented on localization/ndt_scan_matcher/src/ndt_scan_matcher_core.cpp:
```diff
@@ -704,6 +723,72 @@ bool NDTScanMatcher::validate_converged_param(
+  }
+  Eigen::Vector2d centroid = tmp_centroid / (offset_array_.size() + 1);
+  // TODO unbiased covariance: R * ((n-1)/n))
+  Eigen::Matrix2d pca_covariance;
```
This as well in terms of initialization.

kminoda reviewed (commented):
Hi, thank you for the PR :tada: 
Left some minor comments so please check :pray:

YamatoAndo commented:
@KOKIAOKI Thank you for your contribution!
Could you write an explanation in the readme?
I'd also like you to include a link of your paper in the readme.

YamatoAndo commented on localization/ndt_scan_matcher/src/ndt_scan_matcher_core.cpp:
```diff
@@ -704,6 +723,72 @@ bool NDTScanMatcher::validate_converged_param(
+    multi_initial_pose_msg.poses.push_back(matrix4f_to_pose(sub_initial_pose_matrix));
+  }
+  Eigen::Vector2d centroid = tmp_centroid / (offset_array_.size() + 1);
+  // TODO unbiased covariance: R * ((n-1)/n))
```
CI/CD says,
> localization/ndt_scan_matcher/src/ndt_scan_matcher_core.cpp:780:  Missing username in TODO; it should look like "// TODO(my_username): Stuff."  [readability/todo] [2]

KOKIAOKI commented:
Thank you for checking my implementation.
Please let me test again after fixing these fixes.
Also, to pass the DCO, I will rebase past commits and group them together, and add Signed-off to the commit message.

YamatoAndo commented on localization/ndt_scan_matcher/src/ndt_scan_matcher_core.cpp:
```diff
@@ -435,6 +447,13 @@ void NDTScanMatcher::callback_sensor_points(
+  // covariance estimation
+  std::array<double, 36> ndt_covariance;
+  ndt_covariance = output_pose_covariance_;
+  if (is_converged && use_cov_estimation_) {
```
`is_converged` is defined at line 491, so it will cause a compile error.

YamatoAndo commented on localization/ndt_scan_matcher/src/ndt_scan_matcher_core.cpp:
```diff
@@ -704,6 +723,71 @@ bool NDTScanMatcher::validate_converged_param(
+  // Laplace approximation
+  const Eigen::Matrix2d hessian_inv = -ndt_result.hessian.inverse().block(0, 0, 2, 2);
+  const Eigen::SelfAdjointEigenSolver<Eigen::Matrix2d> lap_eigensolver(hessian_inv);
+  Eigen::Matrix2d rot = Eigen::Rotation2Dd(0.0);
```
compile error.

YamatoAndo commented on localization/ndt_scan_matcher/src/ndt_scan_matcher_core.cpp:
```diff
@@ -704,6 +723,71 @@ bool NDTScanMatcher::validate_converged_param(
+  geometry_msgs::msg::PoseArray multi_ndt_result_msg;
+  geometry_msgs::msg::PoseArray multi_initial_pose_msg;
+  multi_ndt_result_msg.header.stamp = sensor_ros_time;
+  multi_initial_pose_msg.header.stamp = sensor_ros_time;
```
```suggestion
  multi_initial_pose_msg.header.stamp = sensor_ros_time;
  multi_ndt_result_msg.header.frame_id = map_frame_;
  multi_initial_pose_msg.header.frame_id = map_frame_;
```
need frame_id to show RViz

KYabuuchi reviewed (commented):
I added some comments. Please check it out :pray:

KYabuuchi commented on localization/ndt_scan_matcher/src/ndt_scan_matcher_core.cpp:
```diff
@@ -466,6 +473,18 @@ void NDTScanMatcher::callback_sensor_points(
   }
 
+  // covariance estimation
+  std::array<double, 36> ndt_covariance;
+  ndt_covariance = output_pose_covariance_;
```
```suggestion
  std::array<double, 36> ndt_covariance = output_pose_covariance_;
```

Motsu-san commented on localization/ndt_scan_matcher/README.md:
```diff
@@ -257,3 +259,21 @@ This is a function that uses de-grounded LiDAR scan to estimate the scan matchin
 | `estimate_scores_for_degrounded_scan` | bool   | Flag for using scan matching score based on de-grounded LiDAR scan (FALSE by default) |
 | `z_margin_for_ground_removal`         | double | Z-value margin for removal ground points                                              |
+
+## 2D rea-time covariance estimation
```
If this function consumes much calculation resource as that may spoil healthy system behavior, could you inform that in this README ?

KYabuuchi commented on localization/ndt_scan_matcher/src/ndt_scan_matcher_core.cpp:
```diff
@@ -692,6 +711,67 @@ bool NDTScanMatcher::validate_converged_param(
+
+  const Eigen::Matrix2d pca_covariance =
+    (tmp_cov / (offset_array_.size() + 1) - (centroid * centroid.transpose()));
+  ndt_covariance[0] = output_pose_covariance_[0] + pca_covariance(0, 0);
+  ndt_covariance[1] = pca_covariance(1, 0);
+  ndt_covariance[6] = pca_covariance(0, 1);
+  ndt_covariance[7] = output_pose_covariance_[7] + pca_covariance(1, 1);
```
This code assumes that the roll/pitch/yaw of `ndt_covariance` is initialized elsewhere. Since this is risking becoming a bug, how about the following?
```
ndt_covariance = output_pose_covariance_;
ndt_covariance[0 + 6 * 0] += pca_covariance(0, 0);
ndt_covariance[1 + 6 * 0] += pca_covariance(1, 0);
ndt_covariance[0 + 6 * 1] += pca_covariance(0, 1);
ndt_covariance[1 + 6 * 1] += pca_covariance(1, 1);
```

KOKIAOKI commented:
Sorry for the delay in reporting.
I have reflected all the reviews made by @Motsu-san  and @KYabuuchi  into the source code.
- Changed the calculation of variance to a method that uses smaller variables.
- There were many variable names that caused misunderstandings, so we made them easier to understand.
- Changed and added notes and explanations in README

YamatoAndo commented on localization/ndt_scan_matcher/README.md:
```diff
@@ -257,3 +259,24 @@ This is a function that uses de-grounded LiDAR scan to estimate the scan matchin
+| -----------------------------| ------------------- | -----------------------------------------------------------------|
+| `use_covariance_estimation`  | bool                | Flag for using real-time covariance estimation (FALSE by default)|
+| `initial_pose_offset_model_x`| std::vector<double> | X-axis offset [m]                                                |
+| `initial_pose_offset_model_y`| std::vector<double> | Y-axis offset [m]                                                |
```
```suggestion
| `initial_pose_offset_model_y`| std::vector<double> | Y-axis offset [m]                                                |

```

by pre-commit.ci
> fix end of files.........................................................Failed
> - hook id: end-of-file-fixer
> - exit code: 1
> - files were modified by this hook
> 
> Fixing localization/ndt_scan_matcher/README.md

KOKIAOKI commented:
@YamatoAndo @KYabuuchi 
Thank you for further checks.
I created the function of eigenvector calculation and added README empty line.

SakodaShintaro commented on localization/localization_util/include/localization_util/util_func.hpp:
```diff
@@ -29,13 +29,20 @@
 #include <string>
 #include <vector>
 
+std::vector<Eigen::Vector2d> create_initial_pose_offset_model(
```
From my personal point of view, `localization_util` should have functions that are likely to be used in other packages.

It seems that `create_initial_pose_offset_model` and `calc_eigenvector_angle` are only used in `ndt_scan_matcher`, and they appear to be called only once, so I think it might not be a bad idea to write them directly at the point of use. 

This is not an absolute necessity for a change, and I welcome any opposing views or comments.

KYabuuchi replied:
I'm sorry. I asked Aoki-san to do so...

SakodaShintaro replied:
OK, no problem.
Let's leave it as it is.

KYabuuchi replied:
Isn't `calc_eigenvector_angle` general purpose? (By the way, the function name is not so good.)

I do not think `create_initial_pose_offset_model` needs to be included in the localizaton_util. Because it is very limited usage.

KOKIAOKI replied:
Two functions, create_initial_pose_offset_model and calc_eigenvector_angle are limited usage.
Can I put these functions in ndt_scan_matcher_core.cpp as **non-member** functions of NDTScanMatcher?

SakodaShintaro commented on localization/ndt_scan_matcher/src/ndt_scan_matcher_core.cpp:
```diff
@@ -141,6 +141,21 @@ NDTScanMatcher::NDTScanMatcher()
+      initial_pose_offset_model_ =
+        create_initial_pose_offset_model(initial_pose_offset_model_x, initial_pose_offset_model_y);
+    } else {
+      use_cov_estimation_ = false;
```
I feel it would be better to make it an explicit error here than to make it false.
I am not sure if the node should be shut down, but I wanted at least a message to the log.

SakodaShintaro commented on localization/localization_util/src/util_func.cpp:
```diff
@@ -18,6 +18,32 @@
+  return initial_pose_offset_model;
+}
+
+Eigen::Matrix2d calc_eigenvector_angle(const Eigen::Matrix2d & matrix)
```
This function returns `Eigen::Matrix2d::Zero()` when `eigensolver.info()` is not Success, which means failure.
I would like to see you use `std::optional` or `exception` so that failure can be more explicitly identified.
If failures occur with a not-insignificant probability, I would suggest using `std::optional`, and if failures rarely occur, I would suggest using `exception`.

KYabuuchi replied:
Also, please change the function name like followings :pray:
* `compute_rotation_matrix_to_decompose_by_pca(const Eigen::Matrix2d & covariance_matrix)`
* `find_rotation_matrix_aligning_covariance_to_principal_axes(const Eigen::Matrix2d & covariance_matrix)`

YamatoAndo commented on localization/ndt_scan_matcher/include/ndt_scan_matcher/ndt_scan_matcher_core.hpp:
```diff
@@ -25,6 +25,7 @@
 
 #include <diagnostic_msgs/msg/diagnostic_array.hpp>
 #include <geometry_msgs/msg/pose_with_covariance_stamped.hpp>
+#include <geometry_msgs/msg/pose_array.hpp>
```
by clang-format
```suggestion
#include <geometry_msgs/msg/pose_array.hpp>
#include <geometry_msgs/msg/pose_with_covariance_stamped.hpp>
```

I think that these files need to be arranged in alphabetical order.

KOKIAOKI commented:
@YamatoAndo @KYabuuchi @SakodaShintaro 
Thank you for further code checks.
Main changes are as below:
- @SakodaShintaro Like your advice, I added parameter warning and std::exeption about Eigen::Success.
- @SakodaShintaro Although review about util.cpp was closed, I moved two functions from util.cpp to ndt_scan_matcher.cpp.

YamatoAndo commented on localization/ndt_scan_matcher/README.md:
```diff
@@ -34,6 +34,8 @@ One optional function is regularization. Please see the regularization chapter i
 | `points_aligned`                  | `sensor_msgs::msg::PointCloud2`                 | [debug topic] pointcloud aligned by scan matching                                                                                        |
 | `points_aligned_no_ground`        | `sensor_msgs::msg::PointCloud2`                 | [debug topic] de-grounded pointcloud aligned by scan matching                                                                            |
 | `initial_pose_with_covariance`    | `geometry_msgs::msg::PoseWithCovarianceStamped` | [debug topic] initial pose used in scan matching                                                                                         |
+| `multi_ndt_pose`                  | `geometry_msgs::msg::PoseArray`                 | [debug topic] estimated poses from multiple initial poses in real-time covariance estimation                                                                                                           |
+| `multi_initial_pose`              | `geometry_msgs::msg::PoseArray`                 | [debug topic] initial poses for real-time covariance estimation                                                                                                           |
```
please delete unnecessary spaces
https://github.com/autowarefoundation/autoware.universe/pull/5454/commits/3744ee3525df05e16678c067af86bcd21e88c552

KOKIAOKI replied:
@YamatoAndo 
I follow the Markdown table style lile in other sections.
Should I remove unnecessary spaces?

YamatoAndo replied:
yes, please remove spaces so that the tables align.

![Screenshot from 2023-11-02 13-42-06](https://github.com/autowarefoundation/autoware.universe/assets/13819566/a6cbb964-b174-4a19-98e7-10235f138133)

KOKIAOKI replied:
Sorry, I was mistaken. I will fix them now.

YamatoAndo commented on localization/ndt_scan_matcher/README.md:
```diff
@@ -257,3 +259,25 @@ This is a function that uses de-grounded LiDAR scan to estimate the scan matchin
+### Parameters
+
+initial_pose_offset_model is rotated around (x,y) = (0,0) in the direction of the first principal component of the Hessian matrix.
+initial_pose_offset_model_x & initial_pose_offset_model_y must have the same number of elements.
+| Name                         | Type                | Description                                                      |
```
I think a line break is necessary.

```suggestion
initial_pose_offset_model_x & initial_pose_offset_model_y must have the same number of elements.

| Name                         | Type                | Description                                                      |
```

YamatoAndo commented on localization/ndt_scan_matcher/README.md:
```diff
@@ -257,3 +259,26 @@ This is a function that uses de-grounded LiDAR scan to estimate the scan matchin
+initial_pose_offset_model is rotated around (x,y) = (0,0) in the direction of the first principal component of the Hessian matrix.
+initial_pose_offset_model_x & initial_pose_offset_model_y must have the same number of elements.
+
+| Name                         | Type                | Description                                                      |
+| -----------------------------| ------------------- | -----------------------------------------------------------------|
+| `use_covariance_estimation`  | bool                | Flag for using real-time covariance estimation (FALSE by default)|
+| `initial_pose_offset_model_x`| std::vector<double> | X-axis offset [m]                                                |
+| `initial_pose_offset_model_y`| std::vector<double> | Y-axis offset [m]                                                |
```
add spaces 
https://github.com/autowarefoundation/autoware.universe/pull/5454/commits/2eea6b020f99ed34e819ca1107e25f96f1ed4ab9
